src/vs/editor/contrib/hover/browser/markdownHoverParticipant.ts

```diff
@@ -120,21 +143,252 @@ export class MarkdownHoverParticipant implements IEditorHoverParticipant<Markdow
+}
+
+type RenderedHoverPart = RenderedNonVerboseHoverPart | RenderedVerboseHoverPart;
+
+interface FocusedHoverInfo {
+	hoverPartIndex: number;
+	// TODO@aiday-mar is this needed?
+	focusRemains: boolean;
+}
+
+class RenderedHoverParts extends Disposable {
+
+	private _renderedHoverParts: RenderedHoverPart[];
+	private _hoverFocusInfo: FocusedHoverInfo = { hoverPartIndex: -1, focusRemains: false };
+
+	constructor(
+		hoverParts: MarkdownHover[],
+		hoverPartsContainer: DocumentFragment,
+		private readonly _editor: ICodeEditor,
+		private readonly _languageService: ILanguageService,
+		private readonly _openerService: IOpenerService,
+		private readonly _keybindingService: IKeybindingService,
+		private readonly _hoverService: IHoverService,
+		private readonly _onFinishedRendering: () => void,
+	) {
+		super();
+		this._renderedHoverParts = this._renderHoverParts(hoverParts, hoverPartsContainer, this._onFinishedRendering);
+		this._register(toDisposable(() => {
+			this._renderedHoverParts.forEach(renderedHoverPart => {
+				renderedHoverPart.disposables.dispose();
+			});
+		}));
+	}
+
+	private _renderHoverParts(
+		hoverParts: MarkdownHover[],
+		hoverPartsContainer: DocumentFragment,
+		onFinishedRendering: () => void,
+	): RenderedHoverPart[] {
+		hoverParts.sort(compareBy(hover => hover.ordinal, numberComparator));
+		return hoverParts.map((hoverPart, hoverIndex) => {
+			const renderedHoverPart = this._renderHoverPart(
+				hoverIndex,
+				hoverPart.contents,
+				hoverPart.source,
+				onFinishedRendering
+			);
+			hoverPartsContainer.appendChild(renderedHoverPart.renderedMarkdown);
+			return renderedHoverPart;
+		});
+	}
+
+	private _renderMarkdownContent(
+		markdownContent: IMarkdownString[],
+		onFinishedRendering: () => void
+	): RenderedBaseHoverPart {
+		const renderedMarkdown = $('div.hover-row');
+		renderedMarkdown.tabIndex = 0;
+		const renderedMarkdownContents = $('div.hover-row-contents');
+		renderedMarkdown.appendChild(renderedMarkdownContents);
+		const disposables = new DisposableStore();
+		disposables.add(renderMarkdownInContainer(
+			this._editor,
+			renderedMarkdownContents,
+			markdownContent,
+			this._languageService,
+			this._openerService,
+			onFinishedRendering,
+		));
+		return { renderedMarkdown, disposables };
+	}
+
+	private _renderHoverPart(
+		hoverPartIndex: number,
+		hoverContents: IMarkdownString[],
+		hoverSourceInformation: HoverSourceInformation | undefined,
+		onFinishedRendering: () => void
+	): RenderedHoverPart {
+
+		const { renderedMarkdown, disposables } = this._renderMarkdownContent(hoverContents, onFinishedRendering);
+
+		if (!hoverSourceInformation) {
+			return { isVerbose: false, renderedMarkdown, disposables };
+		}
+
+		const hover = hoverSourceInformation.hover;
+		const canIncreaseVerbosity = hover.canIncreaseVerbosity;
+		const canDecreaseVerbosity = hover.canDecreaseVerbosity;
+
+		disposables.add(toDisposable(() => {
+			hover.dispose();
+		}));
+
```

Comment:
When you see this, you should ask yourself: Who disposes the hover if there are no `hoverSourceInformation`?